driver/srf02: expose trigger and read to API

haukepetersen opened the pull request:
based on #4782

When using more than one sensor on the same bus, it is very helpful to split the trigger and read functions and do the waiting outside of the driver. This way multiple devices can be triggered and the 70ms waiting period can be spend while all devices are doing their ranging in parallel. With the current implementation we have to wait 70ms for each sensor after each other and so simultaneous measurements triggered from the same thread are not possible...

PeterKietzmann commented:
#4782 was merged. Your approach makes totally sense to me. Might take some days until I can test it

PeterKietzmann commented on tests/driver_srf02/main.c:
```diff
@@ -55,10 +72,46 @@ int main(void)
+    uint8_t new_addr;
+
+    if (argc < 2) {
+        printf("usage: %s <new_addr (decimal)>\n", argv[0]);
```
In the header file you changed the format of `SRF02_DEFAULT_ADDR` to hex but here you ask for decimal numbers. I'd like to unify that

haukepetersen replied:
The problem is rather that I didn't want to implement parsing hex-strings in this simple test application. The devices datasheet is actually using hex addresses, and this is why I changed the addressing in the header file to hex, also...

PeterKietzmann commented on tests/driver_srf02/main.c:
```diff
@@ -28,26 +36,35 @@
-    puts("This test will sample the sensor once per second and display the\n"
-         "result\n");
+    if (argc < 2) {
+        printf("usage: %s <addr (decimal)>\n", argv[0]);
```
In the header file you changed the format of `SRF02_DEFAULT_ADDR` to hex but here you ask for decimal numbers. I'd like to unify that

haukepetersen replied:
actually this because of the simplicity of parsing ints vs hex in the shell...

PeterKietzmann commented:
1. Initializing with `init 224` and starting the sampling loop with `sample` works at the beginning but then always leads to overflows (/or similar). Output is always `distance = 255 cm` after a short time. 

2. How is the work flow for resetting the address? Would be nice to have a hint in the applications documentation. Anyway, I didn't get it working to reprogram the address and sample a value then (well somehow I was, but it returned `distance = 29703 cm`)

3. It might be of interest to have a shell command that just ready the value once

PeterKietzmann commented:
In addition: Just tested the sensor connected to a samr21-xpro

PeterKietzmann commented:
@haukepetersen let's postpone the hex/dec discussion and focus on the functionality. (How) Did you successfully test the driver or did you experience same as described above?

haukepetersen commented:
@PeterKietzmann: updated the documentation in the test and in the README (hope everything is clearer now). Also fixed a bug in the initialization of the driver (missing `i2c_release` on error in init), and finally added a single-shot shell command `shoot`.

Just tested everything again with a `samr21-xpro` - works just fine. Here is some example output:
```
2016-03-21 15:59:02,833 - INFO # main(): This is RIOT! (Version: 2016.03-devel-1271-gfbcec6-prag-opt_srf02_impltrigger)
2016-03-21 15:59:02,834 - INFO # 
2016-03-21 15:59:02,836 - INFO # SRF02 Ultrasonic Range Sensor Test
2016-03-21 15:59:02,836 - INFO # 
2016-03-21 15:59:02,842 - INFO # Use the following flow to test your device/setup. First you need to
2016-03-21 15:59:02,848 - INFO # initialize your device (e.g. 'init 224'). Next you can sample your 
2016-03-21 15:59:02,853 - INFO # device continuously ('sample'), or get one value ('shoot').
2016-03-21 15:59:02,853 - INFO # 
2016-03-21 15:59:02,859 - INFO # This application let's you also reprogram your device's I2Caddress:
2016-03-21 15:59:02,863 - INFO #  1. initialize your device -> e.g. 'init 224'
2016-03-21 15:59:02,867 - INFO #  2. specify the new address -> e.g.'addr 228'
2016-03-21 15:59:02,872 - INFO # The device will be programmed with the new address and it is
2016-03-21 15:59:02,878 - INFO # re-initialized with the new address, so you can use it right away
2016-03-21 15:59:02,878 - INFO # 
init 228
2016-03-21 15:59:05,350 - INFO # > init 228
2016-03-21 15:59:05,354 - INFO # Initializing SRF02 sensor at I2C_DEV(0), address is 228
2016-03-21 15:59:05,379 - INFO # ... [Ok]
2016-03-21 15:59:05,379 - INFO # 
shoot
2016-03-21 15:59:07,784 - INFO # > shoot
2016-03-21 15:59:07,857 - INFO # distance =  94 cm

2016-03-21 15:59:08,173 - INFO # > shoot
2016-03-21 15:59:08,245 - INFO # distance =  94 cm

2016-03-21 15:59:08,459 - INFO # > shoot
2016-03-21 15:59:08,532 - INFO # distance =  94 cm

2016-03-21 15:59:08,659 - INFO # > shoot
2016-03-21 15:59:08,731 - INFO # distance =  94 cm
init 224
2016-03-21 15:59:12,778 - INFO # > init 224
2016-03-21 15:59:12,782 - INFO # Initializing SRF02 sensor at I2C_DEV(0), address is 224
2016-03-21 15:59:12,807 - INFO # ... [Failed]
shoot
2016-03-21 15:59:14,065 - INFO # > shoot
2016-03-21 15:59:14,137 - INFO # distance = 29703 cm

2016-03-21 15:59:14,869 - INFO # > shoot
2016-03-21 15:59:14,941 - INFO # distance = 29703 cm

2016-03-21 15:59:15,054 - INFO # > shoot
2016-03-21 15:59:15,126 - INFO # distance = 29703 cm
init 228
2016-03-21 15:59:17,308 - INFO # > init 228
2016-03-21 15:59:17,313 - INFO # Initializing SRF02 sensor at I2C_DEV(0), address is 228
2016-03-21 15:59:17,337 - INFO # ... [Ok]
2016-03-21 15:59:17,338 - INFO # 
shoot
2016-03-21 15:59:19,078 - INFO # > shoot
2016-03-21 15:59:19,150 - INFO # distance =  94 cm

2016-03-21 15:59:19,799 - INFO # > shoot
2016-03-21 15:59:19,872 - INFO # distance =  94 cm
addr 224
2016-03-21 15:59:22,372 - INFO # > addr 224
2016-03-21 15:59:22,374 - INFO # Set address to 224
shoot
2016-03-21 15:59:23,947 - INFO # > shoot
2016-03-21 15:59:24,019 - INFO # distance =  94 cm

2016-03-21 15:59:24,432 - INFO # > shoot
2016-03-21 15:59:24,504 - INFO # distance =  94 cm
init 228
2016-03-21 15:59:29,054 - INFO # > init 228
2016-03-21 15:59:29,059 - INFO # Initializing SRF02 sensor at I2C_DEV(0), address is 228
2016-03-21 15:59:29,083 - INFO # ... [Failed]
shoot
2016-03-21 15:59:30,125 - INFO # > shoot
2016-03-21 15:59:30,198 - INFO # distance = 29703 cm
init 224
2016-03-21 15:59:31,888 - INFO # > init 224
2016-03-21 15:59:31,893 - INFO # Initializing SRF02 sensor at I2C_DEV(0), address is 224
2016-03-21 15:59:31,917 - INFO # ... [Ok]
2016-03-21 15:59:31,918 - INFO # 
shoot
2016-03-21 15:59:33,047 - INFO # > shoot
2016-03-21 15:59:33,119 - INFO # distance =  94 cm

2016-03-21 15:59:33,529 - INFO # > shoot
2016-03-21 15:59:33,601 - INFO # distance =  94 cm
sample
2016-03-21 15:59:36,073 - INFO # > sample
2016-03-21 15:59:36,146 - INFO # distance =  94 cm
2016-03-21 15:59:36,246 - INFO # distance =  94 cm
2016-03-21 15:59:36,347 - INFO # distance =  94 cm
2016-03-21 15:59:36,447 - INFO # distance =  94 cm
2016-03-21 15:59:36,548 - INFO # distance =  94 cm
2016-03-21 15:59:36,648 - INFO # distance =  94 cm
2016-03-21 15:59:36,749 - INFO # distance =  94 cm
2016-03-21 15:59:36,849 - INFO # distance =  94 cm
2016-03-21 15:59:36,950 - INFO # distance =  94 cm
2016-03-21 15:59:37,050 - INFO # distance =  94 cm
2016-03-21 15:59:37,151 - INFO # distance =  94 cm
2016-03-21 15:59:37,251 - INFO # distance =  94 cm
2016-03-21 15:59:37,352 - INFO # distance =  94 cm
2016-03-21 15:59:37,452 - INFO # distance =  94 cm
2016-03-21 15:59:37,553 - INFO # distance =  94 cm
2016-03-21 15:59:37,653 - INFO # distance =  94 cm
2016-03-21 15:59:37,754 - INFO # distance =  94 cm
2016-03-21 15:59:37,854 - INFO # distance =  94 cm
2016-03-21 15:59:37,955 - INFO # distance =  94 cm
2016-03-21 15:59:38,056 - INFO # distance =  94 cm
2016-03-21 15:59:38,156 - INFO # distance =  94 cm
2016-03-21 15:59:38,257 - INFO # distance =  94 cm
2016-03-21 15:59:38,357 - INFO # distance =  94 cm
2016-03-21 15:59:38,457 - INFO # distance =  94 cm
2016-03-21 15:59:38,558 - INFO # distance =  94 cm
2016-03-21 15:59:38,659 - INFO # distance =  94 cm
2016-03-21 15:59:38,759 - INFO # distance =  94 cm
2016-03-21 15:59:38,859 - INFO # distance =  94 cm
2016-03-21 15:59:38,960 - INFO # distance =  94 cm
2016-03-21 15:59:39,061 - INFO # distance =  94 cm
2016-03-21 15:59:39,161 - INFO # distance =  94 cm
2016-03-21 15:59:39,261 - INFO # distance =  94 cm
2016-03-21 15:59:39,362 - INFO # distance =  94 cm
2016-03-21 15:59:39,463 - INFO # distance =  94 cm
2016-03-21 15:59:39,563 - INFO # distance =  94 cm
...
```

PeterKietzmann commented on drivers/include/srf02.h:
```diff
@@ -34,7 +34,12 @@ extern "C" {
  * @brief   Default I2C address of SRF02 sensors
  */
-#define SRF02_DEFAULT_ADDR                  (0xe0)
+#define SRF02_DEFAULT_ADDR      (0xe0)
```
What do you think about spending a comment e.g. `/* 0d224 */`?

PeterKietzmann commented on tests/driver_srf02/README.md:
```diff
@@ -2,9 +2,49 @@
 
-After initialization, the sensor value is read periodically and printed to the STDOUT.
+This test application also allows to re-program the I2C address of SRF02
+devices. For this initialize the device with its current address and than change
```
th**e**n

PeterKietzmann commented on tests/driver_srf02/README.md:
```diff
@@ -2,9 +2,49 @@
-After initialization, the sensor value is read periodically and printed to the STDOUT.
+This test application also allows to re-program the I2C address of SRF02
+devices. For this initialize the device with its current address and than change
+the to the new address using the `addr` shell command (e.g. `addr 228`). After
```
one "the" too much

PeterKietzmann commented on tests/driver_srf02/README.md:
```diff
@@ -2,9 +2,49 @@
+the to the new address using the `addr` shell command (e.g. `addr 228`). After
+calling the `addr` command, this application will automatically re-initialize
+the device with the new address, so it is usable right away.
 
```
IMO it's a stupid hardware design that you can change the address (non volatile) but can't read it out again (except looking at the LED at startup). It could be useful to describe that quickly or refer to the datasheet

haukepetersen replied:
I think we have to draw a line here, it does not make sense to 'copy' the datasheet here, so if one wants to use the device, one needs to read the datasheet in anyway... I don't see it as out duty to include everything possible here, but to describe briefly what this application does and how it will behave...

PeterKietzmann commented:
Now everything works as expected, nice :-)

haukepetersen commented:
addressed comments, squashed and rebased. Do you ack?

PeterKietzmann commented:
yes okay :-)

PeterKietzmann commented:
and go
